### DISCUSSION TITLE
Zoom using ctrl + middle mouse drag

### DISCUSSION THREAD

**Discussion by DeMastah** [mbrlabs/Lorien#69] [Ideas]
As far as I am aware, the only way to zoom the canvas is by using the scroll wheel. I'd like to be able to zoom by moving the cursor up and down on the screen while pressing <kbd>CTRL</kbd> and the middle mouse button. This way, one can use the stylus to both pan and zoom without having to use the mouse at all (though scrolling to zoom would still be possible, of course). 

I have implemented a rudimentary proof of concept as an addition to `PanZoomCamera.gd`. It works in tandem with the "scroll to zoom" code. At the moment, moving the cursor up zooms in while moving it down zooms out. I suppose this should be configurable in the settings.

One obstacle that I have not addressed is the fact that the current tool should be disabled while zooming is active, or the user could make accidental strokes. A fix to this problem would probably require modifications to other files; it might even make sense to implement this as a separate tool which is invoked whenever <kbd>CTRL</kbd> and MMB are pressed together.

Seeing as this is my first time contributing to a repository (and after having read `contributing.md`), it didn't feel appropriate to create a pull request right away. I hope I'm doing this right.

**Comment by mbrlabs** [comment]
Yes, that is something that bugged me as well (switching between mouse and pen), so that would be a welcome addition. Ctrl+MMB is great! I also have Krita setup this way.

> At the moment, moving the cursor up zooms in while moving it down zooms out. I suppose this should be configurable in the settings.

No, that's a sane default for now. I don't want to add too many settings anyway.

> One obstacle that I have not addressed is the fact that the current tool should be disabled while zooming is active

Yeah, not sure what the best approch here would be at the moment. The current canvas tool certainly needs to be disabled. 

 > Seeing as this is my first time contributing to a repository (and after having read contributing.md), it didn't feel appropriate to create a pull request right away. I hope I'm doing this right.

Yes, generally if it's a new feature it's always better to first ask here or open an issue. That being said, this is a great feature so you can open a PR right away if you want. Even if it's only a draft/prototype...PRs can always be updated incrementally. I'm currently finalizing the v0.4.0 realease, which i want to release in a couple of days. After that's done i can help you out with the PR ;)

**Reply by DeMastah** [reply]
Sounds great. I'll wait with the pull request for a couple days then to avoid any extra work on your end.

Now, I've found a way to further unify the two zooming methods which involves removing the zoom curve. Instead, the zoom level is simply multiplied by some factor when scrolling the mouse wheel. I believe this is how Krita does it (they appear to multiply/divide by sqrt 2). Is the zoom curve critical in any way, or would it be ok to remove it?

**Reply by mbrlabs** [reply]
The reason i added the zoom curve was mainly because the zoom should not grow linearly, because after a couple of zoom steps you can barley see anything changing. But of course you can do a manual caculation as well, as long as it still feels good.

I guess that would also make`PanZoomCamera.set_zoom_level()` obsolete/simpler .
